### ISSUE TITLE
Dynamic Behavior APIs

### ISSUE THREAD

**Issue by EisenbergEffect** [aurelia/aurelia#114]
I'm using this issue to put down some design thoughts on our dynamic behavior APIs. There are APIs that would be used by the `compose` element internally, but that also would be used directly by developers who need to create and apply attributes or elements on the fly.

> Note: No interface or method names are final here. I'm just trying to get some ideas down...

## Idea 1

For convenience, I'm thinking we'll hang these APIs off of the `Aurelia` object. We'll have APIs for elements, attributes and template controllers. Something like this:

```typescript
const au = new Aurelia();
const a = au.createElement(MyCustomElement);
const b = au.createAttribute(MyCustomAttribute);
const c = au.createTemplateController(MyTemplateController, myTemplate);
```

Rather than having these APIs return the element, attribute and template controller directly, I'm thinking they will return something like an `IBehaviorManipulator`. The reason for this is that once you've got an instance of one of these things, what do you do with it? Well, there are a couple of different things I think developers might want to do. The most common is that I think they'd want to graft it into an existing view hierarchy within their app. With vCurrent, developers had some clunky APIs to create things on the fly and then once they created it, they had to manage the entire lifecycle directly. For vNext, I think you should be able to create one of these behaviors and then graft it into the existing view hierarchy so that it can manage the lifecycle for you from that point forward. Here's what that might look like:

```typescript
const ele = au.createElement(MyCustomElement);
const handle = ele.insertBeforeViewNode(document.getElementById('foo'));
```

At some later time, you could remove the behavior, if you want as well:

```typescript
handle.removeFromView();
```

The interfaces should be defined so that you can't call the wrong methods at the wrong time. So, you might have something like this:

```typescript
class Aurelia {
  createElement(elementType: Constructable): IElementManipulator;
}

interface IElementManipulator {
  insertBeforeViewNode(node: INode): IConnectedElement;
  appendToViewNode(node: INode): IConnectedElement;
}

interface IConnectedElement {
  removeFromView(): IElementManipulator;
}
```

I think the manipulator will look a bit different for attributes and template controllers:

```typescript
interface IAttributeManipulator {
  applyToViewNode(node: INode): IConnectedAttribute;
}

interface IConnectedAttribute {
  removeFromView(): IAttributeManipulator;
}
```

Beyond this, I think people will want to add behaviors to the dom and maintain manual control over the lifecycle. So, I think there would be additional methods on the behavior manipulator, perhaps something like this:

```typescript
interface IElementManipulator {
  takeManualControl(): ICustomElement;
}

interface IAttributeManipulator {
  takeManualControl(): ICustomAttribute;
}
```

Should we allow bindings to be set as well?

```typescript
interface IElementManipulator<TElement> {
  setBindings(props: Record<keyof TElement, IBinding>);
}
```

Elements probably also need the ability to set their content:

```typescript
interface IElementManipulator {
  setContent(nodes: INode[]);
}
```

Alternatively, this could be done as part of the create API:

```typescript
au.createElement(MyCustomElement, {
  bindings: { ... },
  content: [ ... ]
}).insertBeforeViewNode(someNode);
```

Does this cover the main scenarios?

## Idea 2

I'm not sure why this didn't occur to me above, but we could mirror the React.createElement API and open up Aurelia to JSX views.

```
au.createElement(MyElement, props, children);
```

I'm not sure what this would return exactly just yet. I'll need to play around with it a bit.

**Comment by fkleuver** [comment]
> Does this cover the main scenarios?

This is *really* hard to anticipate. A dense surface API like this will likely need a number of iterations and is likely to become subject to lots of feature requests in the future.

I would say just implement what comes to mind and don't even bother thinking about performance (we can tackle that later or have the AOT do it), just make the code as clear and easy to understand/change/extend as possible. The sooner we can actually start playing with it the better..

**Comment by EisenbergEffect** [comment]
I don't want to go much farther with upfront design than this. I'm basing this on what I need for `compose` plus what I've seen people try to do with vCurrent (mostly unsuccessfully) over the last few years.

**Comment by davismj** [comment]
> developers who need to create and apply attributes or elements on the fly

What does this mean? Why would a developer want to do this?

**Comment by EisenbergEffect** [comment]
I've got enough to start in on an implementation, but I'd like a sanity check from team members to make sure I'm not way off base.

**Comment by EisenbergEffect** [comment]
@davismj One scenario is in working with 3rd party controls that have their own internal rendering or template system. In this case, the developer may need to get the HTML after it was rendered and then apply some behaviors to it dynamically. The scenarios tend to be pretty advanced.

I need something like this, at least for elements, to make compose work. I'd like to do it in a way that compose is just a thin layer over an API so I'm trying to think through the other things you might want to do with that API. The two main ones are create on the fly and add into an existing view and create on the fly and manage directly.

**Comment by fkleuver** [comment]
As a first impression it does not come across particularly intuitive. I think you're trying to implement the builder pattern where each method returns an interface allowing only the followup methods, but that's not something you want to be doing right from the start. That'll make things really hard to follow and change

**Comment by EisenbergEffect** [comment]
What I'm trying to avoid is enabling someone to remove a view that hasn't been added or add it somewhere else after it's already been added. Implementation was planned that this would all be the same object, just cast as different interfaces but I can just skip that part. But if I do that, I'll have to have internal guards anyways to prevent developers from putting things into an invalid state. So, I was trying to use the interfaces to represent those state transitions.

Anyone have ideas for other APIs. Constraints are things like:

* You have to be able to create the behavior and add it at different times.
* You have to be able to ensure that when you add it the correct lifecycle is guaranteed. This includes the timing of when the element/attribute actually goes into the dom.
* Need to be able to give low-level control for super advanced users.
* Need to allow the behavior to plug into an existing view so that developers don't have to worry about all the lifecycle stuff (including cleanup) if they don't want to.
* Need to be able to manually remove them after adding them (for some advanced scenarios).
* Needs to support `compose` and application startup scenarios (this should replace the "task" stuff in the Aurelia object today).

**Comment by fkleuver** [comment]
>  Implementation was planned that this would all be the same object, just cast as different interfaces but I can just skip that part. But if I do that, I'll have to have internal guards 

I know the idea behind the interface implementations. I didn't mean to not do it (as in have an alternative way to enforce/protect), but rather to not be concerned with it at all just yet. I'm just worried about overthinking things upfront without building something with it first will lead to an API that doesn't fully align with reality

**Comment by davismj** [comment]
> 3rd party controls that have their own internal rendering or template system

Can you give an example of what you are thinking? I have recently implemented a wrapper for [DataTables](https://datatables.net/) and written a [custom tab component similar to this one](https://bl.ocks.org/kevmeister68/96c96a43a9f6e57708afb2c2aaefb904). I could shed some light on what I found in those cases and what would have been ideal in terms of API, but you may have better example cases.

**Comment by fkleuver** [comment]
jQuery UI, materialize components, bootstrap components, kendo UI, that sort of thing?

**Comment by fkleuver** [comment]
I don't know if it fits in the scope here but one thing I have always missed in vCurrent is the ability to drag-and-drop (reorder) items in a repeater. I ended up with a solution that just completely hacked around everything and reassigned a new array on each reorder. Very inefficient. Being able to intercept / take control over the actual appending/removing of DOM nodes would be very helpful in those cases

**Comment by davismj** [comment]
How would you envision doing that?

**Comment by EisenbergEffect** [comment]
@fkleuver I think this could work for that but...I'm guessing people would rather use a repeater and then just some drag/drop attribute and have it "just work" than have to use this set of APIs.

@davismj What simpler APIs might you want elsewhere? What do you feel is missing?

Note: These APIs are probably handy for unit testing. The attribute "assistance" function I wrote for testing our custom attributes made me start thinking about this again. It felt like something that should be built in somehow, rather than buried in our testing infrastructure. But I think it has runtime value as well.

**Comment by davismj** [comment]
This might dovetail with what you're thinking here. If I have a custom element, I should have a callback that allows me to read and manually or automatically process content in the context of that particular custom element.

```
export class MyCustomElement {
  
  prerender(markup: Element) {

    // As opposed to doing au.createElement(MyCustomElement), I'd rather do something like this and let Aurelia doing the heavy lifting for me. No need for a low level API and manual work.
    let dynamicElement = document.createElement('dynamic');

    // Same ewith custom attributes.
    dynamicElement.setAttribute('dynamic-attribute.bind', 'foo');

    // Also, I might want to take incoming information from the template and store it to the element controller on a per-element basis.
    let divs = Array.from(markup.getElementsByTagName('div'));
    this.names = divs.map((div) => div.textContent);

    // MAYBE, I would want to bootstrap my own stuff using low level aurelia APIs,
    // and if so I can return false to say "don't process the content." That sounds 
    // miserable to me personally though. I'd rather manipulate the incoming 
    // template according to Aurelia syntax and then pass it back to Aurelia and 
    // let Aurelia keep going with that.
    // return false; 
  }
}
```

**Comment by EisenbergEffect** [comment]
Isn't that exactly what `processContent` does today?

**Comment by fkleuver** [comment]
> I think also if you try to add a new element using let dynamicElement = document.createElement('dynamic');, it's too late and Aurelia will not process the element as a custom element.

This is why we need the ability to specifically target the template compiler to any arbitrary piece in the DOM, to allow people to dynamically add html and have it work.

**Comment by EisenbergEffect** [comment]
I think @fkleuver and @davismj are talking about the same thing. You want to basically implement the `attaching` hook, which happens after bound properties are available but before the element is added to the DOM, and then you want to dynamically add custom elements and attributes, which become part of the view.

The APIs I'm proposing above would enable that *without the need for the full JIT*. If you want to hand arbitrary HTML nodes or a string of HTML to the framework and get back something that you can merge into the element's view, then you are going to need to use the JIT. I think that is a separate API from what I'm talking about above, though it might return a similar manipulator for the result, since when someone compiles something, they may want to do different things with it: add it into an existing view hierary or manually control it on some out-of-aurelia-control DOM area.

The APIs I'm proposing above are to enable building the `compose` element, assuming all compilation of dynamic elements has taken place and not JIT or compiler is needed.

**Comment by StrahilKazlachev** [comment]
> assuming all compilation of dynamic elements has taken place and no JIT or compiler is needed.
@EisenbergEffect the way I understand this is that `<compose>` will be usable only in AOT, what am I missing?

**Comment by EisenbergEffect** [comment]
`compose` will be usable in both AOT and JIT. The APIs it will use will be runtime APIs which do not require the JIT but if the JIT is present, those APIs will hand off to the JIT compilers when needed.

**Comment by EisenbergEffect** [comment]
I went with Idea 2 from above. PR is available here: https://github.com/aurelia/aurelia/pull/124

**Comment by swalters** [comment]
We dynamically compose cell elements in ag-grid-aurelia and it's a huge hack in vCurrent.  Better dynamic composition is very welcome by me.  I've also had to do some dynamic rendering in my app for different scenarios.

**Comment by EisenbergEffect** [comment]
We've built the `compose` element on top of the `createElement` API, which should itself provide many more applications. We're also pondering enabling async elements on top of `compose`, now that `compose` can pass through arbitrary bindings, due to new capabilities in `createElement`. For example, we could have the compiler transform this code:

```xml
<my-element foo.bind="bar" async><my-element>
```

into this code

```xml
<compose subject.bind="'my-element' | asyncElement" foo.bind="bar"></compose>
```

Then, developers could register async element definitions, maybe something like this:

```TypeScript
aurelia.asyncElement('my-element', () => import('some/package').MyElement);
```

This would be fairly easy to build with the new structures in place.
